### PR TITLE
ECS Executor - add support to adopt orphaned tasks.

### LINKED ISSUE #35491: ECS Executor - Adopting Task Instances
### Description

From feedback [here](https://github.com/apache/airflow/pull/34381#issuecomment-1796224734):

> This is a must-have for us, as our deployments replace our scheduler instances.

> There is a https://github.com/aelzeiny/airflow-aws-executors/pull/15 for that feature in @aelzeiny's executor. I had to make some changes to get that working properly.

### Use case/motivation

_No response_

### Related issues

#34381

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ferruzzi**
Adds support for "adopt orphaned tasks" to the ECS Executor.  

A task can become orphaned, for example, when the scheduler container is re-deployed or terminated or other cases where the SchedulerJob got interrupted.  `try_adopt_task_instances` is defined in the BaseExecutor (this PR implements it in the ECS Executor) and is called by the Scheduler [here](https://github.com/apache/airflow/blob/main/airflow/jobs/scheduler_job_runner.py#L1667).

Based heavily on the [work here](https://github.com/aelzeiny/airflow-aws-executors/pull/15) by Matt Ellis (who I can't tag for some reason) .

closes: https://github.com/apache/airflow/issues/35491

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/executors/ecs/utils.py [hunk 1]
```diff
@@ -46,6 +46,7 @@
     "assign_public_ip": "False",
     "platform_version": "LATEST",
     "check_health_on_startup": "True",
+    "adopt_task_instances": "False",
 }
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/executors/ecs/utils.py:52]: The Changes below are just sorting those dict keys to make it easier to read; no logic changes, just a little housekeeping while I was in there.

File: airflow/providers/amazon/aws/executors/ecs/utils.py [hunk 2]
```diff
@@ -82,22 +83,23 @@ class RunTaskKwargsConfigKeys(BaseConfigKeys):
     ASSIGN_PUBLIC_IP = "assign_public_ip"
     CAPACITY_PROVIDER_STRATEGY = "capacity_provider_strategy"
     CLUSTER = "cluster"
+    CONTAINER_NAME = "container_name"
     LAUNCH_TYPE = "launch_type"
     PLATFORM_VERSION = "platform_version"
     SECURITY_GROUPS = "security_groups"
     SUBNETS = "subnets"
     TASK_DEFINITION = "task_definition"
-    CONTAINER_NAME = "container_name"
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/executors/ecs/utils.py:94]: This change is just sorting the dict keys to make it easier to read; no logic changes, just a little housekeeping while I was in there.

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py [hunk 3]
```diff
@@ -470,3 +481,43 @@ def get_container(self, container_list):
...
+
+            for task in task_descriptions:
+                ti = [ti for ti in tis if ti.external_executor_id == task.task_arn][0]
+                self.active_workers.add_task(
+                    task,
+                    ti.key,
+                    ti.queue,
+                    ti.command_as_list(),
+                    ti.executor_config,
+                    ti.prev_attempted_tries + 1,
```
Comment (eladkal) [inline airflow/providers/amazon/aws/executors/ecs/ecs_executor.py]: What does this +1 means?

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py [hunk 4]
```diff
@@ -470,3 +481,43 @@ def get_container(self, container_list):
             if container["name"] == self.container_name:
                 return container
         raise KeyError(f"No such container found by container name: {self.container_name}")
+
+    def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[TaskInstance]:
+        """
+        Try to adopt running task instances if adopt_task_instances option is set to True.
+
+        These tasks instances should have an ECS process which can be adopted by the unique task ARN.
```
Comment (eladkal) [inline airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:518]: Is it expected that there is no difference for the adoption process between a completed/terminated state to a running state of the instance?

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py [hunk 5]
```diff
@@ -470,3 +481,43 @@ def get_container(self, container_list):
             if container["name"] == self.container_name:
                 return container
         raise KeyError(f"No such container found by container name: {self.container_name}")
+
+    def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[TaskInstance]:
```
Comment (eladkal) [inline airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:514]: may not be part of this PR but just mentioning that for K8s we also have duration metric for the adoption process
which was added in: https://github.com/apache/airflow/pull/35579

I think it would be great to have metric here too

https://github.com/apache/airflow/blob/49108e15eb2eb30e2ccb95c9332db7b38d35f2de/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py#L564

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py]: Should we not be incrementing the tries count?

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py (hunk 5 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:514]: Solid idea.   I didn't even consider adding more metrics here.  I'll give it a think.

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py (hunk 4 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:518]: I think this is right.   For comparison,. k8s executor filters out "status == successful", but local just passes them all, and celery appears to print the state for each but isn't filtering.

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py (hunk 5 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:514]: Added the same.  Thanks for the suggestion.

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py]: Maybe you are right.   It's not re-starting the task, more like reminding the executor that it had to check in on it.... maybe I should drop the +1, I'll look at how the other executors handle it.

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py [hunk 6]
```diff
@@ -377,6 +383,9 @@ def attempt_task_runs(self):
             else:
                 task = run_task_response["tasks"][0]
                 self.active_workers.add_task(task, task_key, queue, cmd, exec_config, attempt_number)
+                # Add Fargate task ARN to executor event buffer, which gets saved
+                # in TaskInstance.external_executor_id.
+                self.event_buffer[task_key] = (State.QUEUED, task.task_arn)
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:404]: We have `self.success` and `self.fail` (e.g. [link](https://github.com/aws-mwaa/upstream-to-airflow/blob/d18c01a7a2f04449262606d79b4f663bdb200414/airflow/executors/base_executor.py#L316)) which handles the other state changes. These are provided by the base executor. I wonder if we should add a new method for putting a task in queued state :thinking: but this might cause weird issues if a provider which contains an executor is installed alongside an older version of airflow... (also, not required for this PR, just thinking out loud)

File: airflow/providers/amazon/provider.yaml [hunk 7]
```diff
@@ -936,6 +936,17 @@ config:
         type: boolean
         example: "True"
         default: "True"
+      adopt_task_instances:
+        description: |
+          If True, the executor will try to adopt orphaned task instances from a SchedulerJob
+          shutdown event, for example when a scheduler container is re-deployed or terminated.
+          If False, the executor will terminate all active AWS Batch Jobs when the scheduler shuts down.
```
Comment (o-nikolas) [inline airflow/providers/amazon/provider.yaml]:
```suggestion
          If False, the executor will terminate all active AWS ECS Tasks when the scheduler shuts down.
```

File: airflow/jobs/scheduler_job_runner.py [hunk 8]
```diff
@@ -721,6 +721,7 @@ def _process_executor_events(self, session: Session) -> int:
             if state == TaskInstanceState.QUEUED:
                 ti.external_executor_id = info
                 self.log.info("Setting external_id for %s to %s", ti, info)
+                session.commit()
```
Comment (o-nikolas) [inline airflow/jobs/scheduler_job_runner.py]: I'm not a DB/SQLA expert, but do we want to commit here, our outside of the for loop? Here is going to do a commit for each queued task individually that we got from the query. Which could be many? Interested to hear others weigh in.

Comment (ferruzzi) [reply airflow/jobs/scheduler_job_runner.py]: A valid point, and I don't know the right answer.   I'll see what others think.  It may be safest to move it outside the loop

**Comment by ferruzzi** [comment]
Accidentally did a rebase and ingested the bug fixed by https://github.com/apache/airflow/pull/36958, re-rebased so CI should clear now.

File: airflow/providers/amazon/aws/executors/ecs/ecs_executor.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py]: @o-nikolas thoughts on this one?

Comment (o-nikolas) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py]: My gut says that we should not be incrementing this. It's just resuming a previous try IMHO. But I'm also curious to see what you dig up when you have a look at other executors.

Comment (ferruzzi) [reply airflow/providers/amazon/aws/executors/ecs/ecs_executor.py]: I'm not seeing it elsewhere.  Removed it.

File: airflow/jobs/scheduler_job_runner.py (hunk 8 above)
Comment (ferruzzi) [reply airflow/jobs/scheduler_job_runner.py]: For posterity, I have started a Slack discussion [[here](https://apache-airflow.slack.com/archives/CCPRP7943/p1706126620502059)] for this to try to get more attention.  I'll drop a comment here with whatever decision is reached there.

**Comment by ferruzzi** [comment]
Unless anyone here feels strongly otherwise, I think I'm going to close this one and open a new PR.  It's been a while and it needs a rebase, plus there are now a few commits which can be squashed, and it looks like it is going to be dependent on a bugfix that I'll submit in a separate PR.  No real point leaving this open and then doing a rebase.
